deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/CudaGradientsAccumulator.java

```diff
@@ -107,16 +159,26 @@ public void applyUpdate(StepFunction function, INDArray params, INDArray updates
         // nullify given updates first
         updates.assign(0.0f);
 
+        int cnt = 0;
         while (!messages.get(index.get()).isEmpty()) {
             INDArray compressed = messages.get(index.get()).poll();
 
-            //log.info("Thread: {}; Compressed: {}", Thread.currentThread().getId(), Arrays.toString(compressed.data().asInt()));
-
-            // FIXME: pass gradients here, and reuse them
             INDArray decoded = Nd4j.getExecutioner().thresholdDecode(compressed, updates);
+            cnt++;
         }
 
-        function.step(params, updates);
+        if (externalSource != null)
+            while (!externalSource.isEmpty()){
+                INDArray compressed = externalSource.poll();
+
+                INDArray decoded = Nd4j.getExecutioner().thresholdDecode(compressed, updates);
+                cnt++;
+            }
+
+        // TODO: average updates probably?
```

Comment:
OK, I'm not sure I understand everything here, so bear with me : is there or isn't there an averaging in the step function ?

At the end of the day, IIUC your accumulation is large / unlimited, and based on that alone it sound like it might diverge. If all m executors (thread, what have you) starts from the same vector wold, then after processing a large batch, the new vector on each executor will be close to the optimal solution w*.

If the variable is updated by vnew = vold + Σ_{i=1}^m ∆(S_i) ,
then we have
wnew − w* =  wold - w* + + Σ_{i=1}^m ∆(S_i) ≈  wold - w* + + Σ_{i=1}^m ∆(w* - wold) = (m-1) (w* - wold) 

Which diverges if m >=3.